File: frontend/src/utils/SqlAnalyzer.ts
/*
 * Copyright 2025 agwlvssainokuni
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * SQL Analysis utilities for frontend pagination support
 * Simplified version of backend SqlAnalyzer for basic ORDER BY detection
 */

export const PagingCompatibility = {
  COMPATIBLE: 'COMPATIBLE',
  NOT_SELECT: 'NOT_SELECT',
  HAS_LIMIT_OFFSET: 'HAS_LIMIT_OFFSET',
  NO_ORDER_BY: 'NO_ORDER_BY'
} as const

export type PagingCompatibility = typeof PagingCompatibility[keyof typeof PagingCompatibility]

/**
 * Check if SQL query is a SELECT statement
 */
export function isSelectQuery(sql: string): boolean {
  const trimmed = sql.trim().toLowerCase()
  return trimmed.startsWith('select') || 
         trimmed.startsWith('with') ||
         trimmed.startsWith('show') ||
         trimmed.startsWith('describe') ||
         trimmed.startsWith('desc') ||
         trimmed.startsWith('explain')
}

/**
 * Check if SQL contains ORDER BY clause
 * Simple implementation - not as sophisticated as backend version
 */
export function hasOrderByClause(sql: string): boolean {
  const normalized = sql.toLowerCase().replace(/\s+/g, ' ')
  return /\border\s+by\b/.test(normalized)
}

/**
 * Check if SQL contains LIMIT clause
 */
export function hasLimitClause(sql: string): boolean {
  const normalized = sql.toLowerCase().replace(/\s+/g, ' ')
  return /\blimit\s+\d+/.test(normalized)
}

/**
 * Check if SQL contains OFFSET clause
 */
export function hasOffsetClause(sql: string): boolean {
  const normalized = sql.toLowerCase().replace(/\s+/g, ' ')
  return /\boffset\s+\d+/.test(normalized)
}

/**
 * Check if SQL has pagination conflicts (LIMIT/OFFSET)
 */
export function hasPagingConflict(sql: string): boolean {
  return hasLimitClause(sql) || hasOffsetClause(sql)
}

/**
 * Get pagination compatibility for SQL query
 */
export function getPagingCompatibility(sql: string): PagingCompatibility {
  if (!isSelectQuery(sql)) {
    return PagingCompatibility.NOT_SELECT
  }
  
  if (hasPagingConflict(sql)) {
    return PagingCompatibility.HAS_LIMIT_OFFSET
  }
  
  if (!hasOrderByClause(sql)) {
    return PagingCompatibility.NO_ORDER_BY
  }
  
  return PagingCompatibility.COMPATIBLE
}

File: src/main/java/cherry/sqlapp2/util/SqlAnalyzer.java
/*
 * Copyright 2025 agwlvssainokuni
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package cherry.sqlapp2.util;

import java.util.regex.Pattern;

/**
 * SQLの解析機能を提供するユーティリティクラス。
 * ページング機能とクエリ構造検出のためのSQL解析を行います。
 * SELECT文の検証、LIMIT/OFFSET句の検出、ORDER BY句の確認などを提供します。
 */
public class SqlAnalyzer {

    // Regex patterns for SQL clause detection (case-insensitive)
    private static final Pattern LIMIT_PATTERN = Pattern.compile(
            "\\blimit\\s+\\d+", 
            Pattern.CASE_INSENSITIVE
    );
    
    private static final Pattern OFFSET_PATTERN = Pattern.compile(
            "\\boffset\\s+\\d+", 
            Pattern.CASE_INSENSITIVE
    );
    
    private static final Pattern ORDER_BY_PATTERN = Pattern.compile(
            "\\border\\s+by\\s+", 
            Pattern.CASE_INSENSITIVE
    );
    
    private static final Pattern SELECT_PATTERN = Pattern.compile(
            "^\\s*select\\s+", 
            Pattern.CASE_INSENSITIVE
    );

    /**
     * SQLにLIMIT句が含まれているかをチェックします。
     * 
     * @param sql チェック対象のSQL文
     * @return LIMIT句が含まれている場合true
     */
    public static boolean hasLimitClause(String sql) {
        if (sql == null || sql.trim().isEmpty()) {
            return false;
        }
        return LIMIT_PATTERN.matcher(sql).find();
    }

    /**
     * SQLにOFFSET句が含まれているかをチェックします。
     * 
     * @param sql チェック対象のSQL文
     * @return OFFSET句が含まれている場合true
     */
    public static boolean hasOffsetClause(String sql) {
        if (sql == null || sql.trim().isEmpty()) {
            return false;
        }
        return OFFSET_PATTERN.matcher(sql).find();
    }

    /**
     * SQLにページング競合（LIMITまたはOFFSET）があるかをチェックします。
     * 
     * @param sql チェック対象のSQL文
     * @return ページング競合がある場合true
     */
    public static boolean hasPagingConflict(String sql) {
        return hasLimitClause(sql) || hasOffsetClause(sql);
    }

    /**
     * Check if SQL contains ORDER BY clause (recommended for pagination)
     * SQLにORDER BY句が含まれているかチェック（ページングに推奨）
     */
    public static boolean hasOrderByClause(String sql) {
        if (sql == null || sql.trim().isEmpty()) {
            return false;
        }
        return ORDER_BY_PATTERN.matcher(sql).find();
    }

    /**
     * SQLがSELECTクエリかをチェックします。
     * SELECT、WITH、SHOW、DESCRIBE、EXPLAIN文を検出します。
     * 
     * @param sql チェック対象のSQL文
     * @return SELECTクエリの場合true
     */
    public static boolean isSelectQuery(String sql) {
        if (sql == null || sql.trim().isEmpty()) {
            return false;
        }
        
        String trimmedSql = sql.trim().toLowerCase();
        return trimmedSql.startsWith("select") ||
               trimmedSql.startsWith("with") ||
               trimmedSql.startsWith("show") ||
               trimmedSql.startsWith("describe") ||
               trimmedSql.startsWith("desc") ||
               trimmedSql.startsWith("explain");
    }

    /**
     * Check if SQL is suitable for pagination
     * SQLがページングに適しているかチェック
     */
    public static boolean isSuitableForPaging(String sql) {
        return isSelectQuery(sql) && 
               !hasPagingConflict(sql) && 
               hasOrderByClause(sql);
    }

    /**
     * Get pagination compatibility status
     * ページング互換性ステータス取得
     */
    public static PagingCompatibility getPagingCompatibility(String sql) {
        if (!isSelectQuery(sql)) {
            return PagingCompatibility.NOT_SELECT;
        }
        
        if (hasPagingConflict(sql)) {
            return PagingCompatibility.HAS_LIMIT_OFFSET;
        }
        
        if (!hasOrderByClause(sql)) {
            return PagingCompatibility.NO_ORDER_BY;
        }
        
        return PagingCompatibility.COMPATIBLE;
    }

    /**
     * Pagination compatibility status enum
     * ページング互換性ステータス列挙型
     */
    public enum PagingCompatibility {
        COMPATIBLE("Compatible with pagination"),
        NOT_SELECT("Not a SELECT query"),
        HAS_LIMIT_OFFSET("Already contains LIMIT/OFFSET clause"),
        NO_ORDER_BY("No ORDER BY clause (results may be inconsistent)");

        private final String description;

        PagingCompatibility(String description) {
            this.description = description;
        }

        public String getDescription() {
            return description;
        }

        public boolean isCompatible() {
            return this == COMPATIBLE;
        }

        public boolean allowsWarning() {
            return this == NO_ORDER_BY;
        }
    }
}